File: android-native/KyouNoOgatore/app/src/main/java/jp/ogatore/kyouno/card/CardData.kt
package jp.ogatore.kyouno.card

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// card-data.json(scripts-native/gen-card-data.mjsがindex.html/app-card.jsから生成。マスタープラン§2-4)の
// Serializableモデル。フィールド名はJS側の定数名(SCREAMING_SNAKE_CASE)を意図的にそのまま使う——
// このデータは「人が読み書きするコード」ではなく「Web版の定数テーブルの1:1写像」であり、9月の
// 差分同期(§5-2)でJS側の変更を機械的に追いやすくすることを、Kotlinの命名規約より優先している。

@Serializable
data class CardTheme(val name: String, val bg: List<String>, val main: String, val deco: List<String>)

@Serializable
data class MilestoneInfo(val d: Int, val t: String)

@Serializable
data class CharaFile(val file: String, val w: Int)

@Serializable
data class SeasonCard(val key: String, val name: String, val bg: List<String>, val ws: Int? = null, val we: Int? = null)

@Serializable
data class RareCard(val key: String, val name: String, val bg: List<String>)

@Serializable
data class NormalCard(val name: String, val bg: List<String>, val main: String, val deco: List<String>)

@Serializable
data class TokuCard(val key: String, val name: String, val bg: List<String>)

@Serializable
data class CardData(
    val CARD_THEMES: List<CardTheme>,
    val GOLD: CardTheme,
    val MS: List<MilestoneInfo>,
    val CHARA_FILES: List<CharaFile>,
    val CARD_THEMES_V1_COUNT: Int,
    val CARD_THEMES_V2_FROM: Int,
    val MILESTONES: List<Int>,
    val CARD_IMG_FROM: Int,
    val SEASON_CARDS: List<SeasonCard>,
    val RARE_CARDS: List<RareCard>,
    val NORMAL_CARDS: List<NormalCard>,
    val TOKU_CARDS: Map<String, TokuCard>, // キーはeffTotalの文字列("4".."1900")
    val CARD_ROT_ORDER: List<Int>,
    // Step 7a(図鑑UI): index.html getDexStatus()が参照するヒント/フレーバー文言。
    val DEX_TEASE: Map<String, String> = emptyMap(), // rareのkey→ロック中ヒント
    val DEX_FLAVOR: Map<String, String> = emptyMap(), // toku/season/rareのkey→アンロック後フレーバー
    val DEX_FLAVOR_NORMAL: Map<String, String> = emptyMap(), // normalのname→アンロック後フレーバー
    val DEX_NORMAL_TEASE: String = "",
)

object CardDataLoader {
    val shared: CardData by lazy {
        val stream = CardDataLoader::class.java.classLoader?.getResourceAsStream("card-data.json")
            ?: error("card-data.json がクラスパスに見つからない(app/src/main/resourcesの同梱漏れ)")
        val text = stream.bufferedReader(Charsets.UTF_8).use { it.readText() }
        Json { ignoreUnknownKeys = true }.decodeFromString(CardData.serializer(), text)
    }
}

File: ios-native/KyouNoOgatore/CardCore/Sources/CardCore/CardData.swift
import Foundation

// card-data.json(scripts-native/gen-card-data.mjsがindex.html/app-card.jsから生成。マスタープラン§2-4)の
// Decodableモデル。フィールド名はJS側の定数名(SCREAMING_SNAKE_CASE)を意図的にそのまま使う——
// このデータは「人が読み書きするコード」ではなく「Web版の定数テーブルの1:1写像」であり、9月の
// 差分同期(§5-2)でJS側の変更を機械的に追いやすくすることを、Swiftの命名規約より優先している。
public struct CardTheme: Decodable {
    public let name: String
    public let bg: [String]
    public let main: String
    public let deco: [String]
}

// UI/UXパリティ監査GO-1(2026-07-28): m(祝いメッセージ)/q(先輩の声引用)は元々d/tだけに
// 間引かれてスキップされていた(節目のお祝い演出が通常日と見分けつかなくなっていた欠落の
// 原因)。gen-card-data.mjsの機械抽出をd/t/m/qの4項目に広げたので、ここも合わせる。
public struct MilestoneInfo: Decodable {
    public let d: Int
    public let t: String
    public let m: String
    public let q: String
}

public struct CharaFile: Decodable {
    public let file: String
    public let w: Int
}

public struct SeasonCard: Decodable {
    public let key: String
    public let name: String
    public let bg: [String]
    public let ws: Int?
    public let we: Int?
}

public struct RareCard: Decodable {
    public let key: String
    public let name: String
    public let bg: [String]
}

public struct NormalCard: Decodable {
    public let name: String
    public let bg: [String]
    public let main: String
    public let deco: [String]
}

public struct TokuCard: Decodable {
    public let key: String
    public let name: String
    public let bg: [String]
}

public struct CardData: Decodable {
    public let CARD_THEMES: [CardTheme]
    public let GOLD: CardTheme
    public let MS: [MilestoneInfo]
    public let CHARA_FILES: [CharaFile]
    public let CARD_THEMES_V1_COUNT: Int
    public let CARD_THEMES_V2_FROM: Int
    public let MILESTONES: [Int]
    public let CARD_IMG_FROM: Int
    public let MILESTONE_MSG_VIDEO: String
    public let SEASON_CARDS: [SeasonCard]
    public let RARE_CARDS: [RareCard]
    public let NORMAL_CARDS: [NormalCard]
    public let TOKU_CARDS: [String: TokuCard] // キーはeffTotalの文字列("4".."1900")
    public let CARD_ROT_ORDER: [Int]
    // Step 7a(図鑑UI): index.html getDexStatus()が参照するヒント/フレーバー文言。
    public let DEX_TEASE: [String: String] // rareのkey→ロック中ヒント
    public let DEX_FLAVOR: [String: String] // toku/season/rareのkey→アンロック後フレーバー
    public let DEX_FLAVOR_NORMAL: [String: String] // normalのname→アンロック後フレーバー
    public let DEX_NORMAL_TEASE: String
}

public enum CardDataLoader {
    public static let shared: CardData = {
        guard let url = Bundle.module.url(forResource: "card-data", withExtension: "json") else {
            fatalError("card-data.json がCardCoreパッケージに同梱されていない")
        }
        do {
            return try JSONDecoder().decode(CardData.self, from: try Data(contentsOf: url))
        } catch {
            fatalError("card-data.json のデコードに失敗: \(error)")
        }
    }()
}
